src/security/PeerConnector.cc

```diff
@@ -543,13 +550,37 @@ Security::PeerConnector::bail(ErrorState *error)
     Must(dialer);
     dialer->answer().error = error;
 
+    if (auto *p = serverConnection()->getPeer())
+        peerConnectFailed(p);
+
     callBack();
-    // Our job is done. The callabck recepient will probably close the failed
-    // peer connection and try another peer or go direct (if possible). We
-    // can close the connection ourselves (our error notification would reach
-    // the recepient before the fd-closure notification), but we would rather
-    // minimize the number of fd-closure notifications and let the recepient
-    // manage the TCP state of the connection.
+    disconnect();
+
+    if (noteFwdPconnUse)
+        fwdPconnPool->noteUses(fd_table[serverConn->fd].pconn.uses);
+    serverConn->close();
+    serverConn = nullptr;
+}
+
+void
+Security::PeerConnector::sendSuccess()
+{
+    callBack();
+    disconnect();
+}
+
+void
+Security::PeerConnector::disconnect()
+{
+    if (!Comm::IsConnOpen(serverConnection()))
```

Comment:
> This check is wrong (see Tunneler::disconnect() for details). AFAICT, 06f1381 (which added this check) describes a real bug (even though the commit message refers to a non-existent PeerConector::retryOrBail() method), 

Sorry I mean PeerConnector::bail() method.

> but the root cause of that bug is not inside disconnect(), and the correct solution for that bug is different.

I decided to make this fix (commit 06f1381), because of two reasons:
- Inside Security::PeerConnector::start there is the following
`
if (!Comm::IsConnOpen(serverConn) || fd_table[serverConn->fd].closing())
        return bail(new ErrorState(ERR_CONNECT_FAIL, Http::scBadGateway, request.getRaw(), al));
`
This is implies that the serverConn maybe be closed for an external reason.
- Having in my mind the possibility where an external user closes the Comm::Connection object we are owning, setting the Comm::Connection::fd to -1. T

Both of the above are only possible if we have a bug, because the PeerConnector is the only owner of Comm::Connection object. For example I entered this state forcing an exception after PeerConnector created but before AsyncJob::Start is called, where in practice there is not any code which can cause an exception. We can have the fd_table[serverConn->fd].closing() but (hopefully) not closed the Comm::Connection.

But because there is the check Comm::IsConnOpen in PeerConnector::start I made the check inside disconnect before use the comm_remove_close_handler for consistency and protect squid from bugs in other places.

> I do not know whether the following change will be enough to address the root cause, but PeerConnector::connectionClosed (i.e. the connection closure callback) must nullify closeHandler (and possibly the server connection itself!).

It does it. But uses the PeerConnector::bail (which calls disconnect) to do it.
Probably we can do it with a cleaner way however requires many changes in current code.
It is better to not try it now, but if you think this is required I can try.

> Removing PeerConnector::connectionClosed(), moving its functionality inside PeerConnector::commCloseHandler() is probably a good idea as well -- why spread the same functionality across two methods

commit 2a53a4663

Reply:
> I decided to make this fix (commit 06f1381), because of two reasons:

> Security::PeerConnector::start ... implies that the serverConn maybe be closed for an external reason.

Yes, it can be. External connection closures is one of the primary reasons why Squid still has a connection closure handler concept (and why this PR adds a start() check).

> Having in my mind the possibility where an external user closes the Comm::Connection object we are owning,

The external connection closure is possible but ...

>  setting the Comm::Connection::fd to -1.

Assume this part is impossible. We are the only Connection owner. Only we can initiate setting Comm::Connection::fd to -1. Without that assumption, it is virtually impossible to write reasonable Connection-handling job code. If you know of any specific place that violates this assumption for PeerConnector, please let me know.

> Both of the above are only possible if we have a bug

The first situation (i.e. external connection closure) is actually possible in some contexts today. One example is the shutdown sequence that initiates closures for connections that shutdown code does not own. Another example is ModPoll's DoSelect(). There may be more. I do not think PeerConnector is directly exposed to the shutdown problem _today_, but, as long as such external closures exist in Squid, I think it is best to assume that PeerConnector (and any other Connection-owning job) is or will be exposed to an external closure.

> because there is the check Comm::IsConnOpen in PeerConnector::start I made the check inside disconnect for consistency and protect squid from bugs in other places.

Consistency is good, but you cannot apply it to the `if` statement condition alone. You have to extend the principle to the code that handles the condition: The start() method does not just quit after detecting an external connection closure! Instead, start() correctly proceeds with what the PeerConnector job is supposed to do (in this case, it means returning an error to the job initiator). Similarly, disconnect() should not just quit -- it should complete what it was asked to do (i.e., prevent any Comm callbacks from reaching the job). Currently, disconnect() violates its promise to disconnect.

I suspect that if you fix that disconnect() violation, then the condition inside that method may itself disappear (or change), but that is a secondary speculation.

> Alex: PeerConnector::connectionClosed must nullify closeHandler.
> Christos: It does it [via] PeerConnector::bail() which calls disconnect().

You are missing the point: disconnect() _removes_ an active closure handler _registration_ (if any). That is _not_ what the closure handler code must do! The closure handler code must clear the handler member (i.e. make closeHandler nil). That is the only way to restore the correct job state. This restoration must happen before calling virtually any other job method because other methods usually assume that the job state is correct. It would be insanely difficult to write correct code if internal methods cannot rely on the job state being valid.

It is possible, even likely, that the closure handler must do _more_ (after syncing the job state with the external changes), of course, but I am focusing on this single critical aspect for now.

> Probably we can do it with a cleaner way however requires many changes in current code.

I doubt the latter is true, but we have to do the right thing here regardless of the cost. Not nullifying closeHandler in a closure handler method is a bug. Fixing that bug is in this PR scope -- this PR works (in part) on correct handling of Connection ownership and connection closure handler is an important part of that ownership lifetime.

Reply:
> >  Security::PeerConnector::start ... implies that the serverConn maybe be closed for an external reason.
 
> Yes, it can be. External connection closures is one of the primary reasons why Squid still has a connection closure handler concept (and why this PR adds a start() check).

> >    Having in my mind the possibility where an external user closes the Comm::Connection object we are owning,

> The external connection closure is possible but ...

 >>  setting the Comm::Connection::fd to -1.

> Assume this part is impossible. We are the only Connection owner. Only we can initiate setting Comm::Connection::fd to -1.

Saying that the "Comm::Connection is closed" or "external user closes the Comm::Connection object" I mean that the Comm::Connection::close is called for this object which sets the Connection::fd to -1. The Comm::IsConnOpen actually  only checks if the Connection::fd is not -1.
This is why I am saying that the Comm::Connection may "closing" (because an external user closed the fd, and we can expect the comm close handler clear us), but is difficult to be closed (have the Comm::Connection::fd to -1)

> Without that assumption, it is virtually impossible to write reasonable Connection-handling job code. If you know of any specific place that violates this assumption for PeerConnector, please let me know.

I agree but we had such bugs in the past. This patch between others also fixes a such bug, because the Comm::Connection object shared with FwdStateData which closed the connection object.
Currently I do not know any specific place which can violate this assumption. But we may have it in the future if an exception thrown after the PeerConnector object is created but before the AsyncJob::Start is executed.  Because of a bug.

> I doubt the latter is true, but we have to do the right thing here regardless of the cost. Not nullifying closeHandler in a closure handler method is a bug. Fixing that bug is in this PR scope -- this PR works (in part) on correct handling of Connection ownership and connection closure handler is an important part of that ownership lifetime.

OK I will try to fix this.

Reply:
Thank you for detailing your terminology. I hope we are on the same page now. I am looking forward to your next update.

Reply:
I commited the patches e9d742e9 a5277e0 daec389f and db3e4ea35 which are related to this.

To answer on Alex questions:

> I do not know whether the following change will be enough to address the root cause, but PeerConnector::connectionClosed (i.e. the connection closure callback) must nullify closeHandler (and possibly the server connection itself!).

This is not enough, we are going to have crash inside PeerConnector::disconnect while the commUnsetConnTimeout be called.

> If necessary, we can remember the timeout callback and then nullify that as well, in PeerConnector::connectionClosed() and in PeerConnector::commTimeoutHandler() callbacks.

I did not do that. Doing this will not resulted to a better code. To avoid code duplication (in commCloseHandler and commTimeoutHandler) a new method required which will looked very similar with disconnect method. The PeerConnector::bail method which is called by commCloseHandler, commTimeoutHandler  and on errors should do the required work without any problem.

Before proceed I must recall that closing the fd is different than closing the Comm::Connection object. An fd can be closing but its Comm::Connection object still considered as open, until the related commClose handlers are called.

My major problem is who can close the server Comm::Connection object. There are classes like the ConnStateData class where its Comm::Connection object can be closed by anyone inside squid and we are expecting that the commClose handler will clear its state. To handle this we have many checks into this class for Comm::IsConnOpen(clientConnection) to check if an external closed our Comm::Connection object.

For the Security::PeerConnector and the Http::Tunneler, thinks looks simpler we have not to do this. We can assume that we are the only owners of our Comm::Connection object and this object can be closed only  by us because of an error or because the commClose handler is called.
Internally we should always call PeerConnector::bail and Http::Tunneler::bailWith methods to clear our state (remove handlers, respond to callers etc) and close the Comm::Connection object.  Moreover these methods allow/require sending an error/reason for the closed connection to the caller, which in practice is required for these classes.

Please also read the comments in my commits.

Reply:
> I must recall that closing the fd is different than closing the Comm::Connection object

Yes, this unfortunate but important fact is probably the source of several bugs. I appreciate you reminding me of that fact and, AFAICT, I agree with your logic surrounding Connection ownership. I made three untested polishing commits on top of your changes: f63ce51, 625f70a, and e61ac40. If you are OK with them, we can resolve this change request.

Reply:
Just the commit 625f70a is not enough to fix squid crashes. Unfortunately the
PeekingPeerConnector::clientConn can be closed, while the request->clientConnectionManager is still valid, for example after a (lifetime?) timeout.

I make the commit 7471528 which replaces the assertion with an "if" check and also add a similar check inside  PeekingPeerConnector::noteNegotiationDone where the switchToTunnel is called which assumes that the clientConn is open.
I hope this is ok.

Reply:
I am OK with your 7471528 changes in principle and am resolving this old change request. The next review will focus on the remaining/current problem.